### PR TITLE
EVG-2964: fix thread levels graphs for items with missing levels

### PR DESCRIPTION
**Pull request by jimoleary**

### CODE REVIEW

**Comment by unrealsolver** [comment]
How does it actually works? As I can see your code just changes `undefined` to `null`, when there are no data.

**Comment by jimoleary** [comment]
It stops exceptions from being thrown.

**Comment by unrealsolver** [comment]
Do `item` is undefined sometimes? If so, maybe it's better to use simpler assertion. If not, could you please give me steps to reproduce?

**Comment by jimoleary** [comment]
Item is always defined but a thread level can be missing.

See [this task](https://evergreen.mongodb.com/task/sys_perf_linux_standalone_core_workloads_WT_ffde025012ca20c0376a22e3a0f52c1887d03a70_18_03_10_13_20_54).

If you click 'All' thread levels,  then you can see the following backtrace in the console:

```javascript
trend_chart.js:134 Uncaught TypeError: Cannot read property 'ops_per_sec' of undefined
    at trend_chart.js:134
    at Function.m.map.m.collect (underscore-min.js:5)
    at SVGPathElement.d (trend_chart.js:417)
    at SVGPathElement.o (d3.min.js:1)
    at d3.min.js:3
    at H (d3.min.js:1)
    at Array.ka.each (d3.min.js:3)
    at Array.ka.attr (d3.min.js:3)
    at drawSingleTrendChart (trend_chart.js:415)
    at drawTrendGraph (perf.js:709)
(anonymous) @ trend_chart.js:134
m.map.m.collect @ underscore-min.js:5
d @ trend_chart.js:417
o @ d3.min.js:1
(anonymous) @ d3.min.js:3
H @ d3.min.js:1
ka.each @ d3.min.js:3
ka.attr @ d3.min.js:3
drawSingleTrendChart @ trend_chart.js:415
drawTrendGraph @ perf.js:709
(anonymous) @ perf.js:425
setTimeout (async)
PerfController.$scope.redrawGraphs @ perf.js:424
(anonymous) @ perf.js:148
$digest @ angular.js:18261
$apply @ angular.js:18531
$$debounceViewValueCommit @ angular.js:29244
$setViewValue @ angular.js:29222
(anonymous) @ angular.js:25672
jg @ angular.js:3771
d @ angular.js:3759
0 Failed to load resource: the server responded with a status of 500 (Internal Server Error)
```

File: public/static/plugins/perf/js/trend_chart.js [hunk 1]
```diff
@@ -122,7 +122,7 @@ mciModule.factory('PerfChartService', function() {
   // thread `level` (ignored in this case) and series `item`
   function getValueForMaxOnly() {
     return function(item) {
-      return item[cfg.valueAttr]
+      return item.hasOwnProperty(cfg.valueAttr) ? item[cfg.valueAttr] : null
```
Comment (unrealsolver) [inline public/static/plugins/perf/js/trend_chart.js]: I don't think this change required in this case.

File: public/static/plugins/perf/js/trend_chart.js [hunk 2]
```diff
@@ -131,7 +131,7 @@ mciModule.factory('PerfChartService', function() {
   // thread `level` and series `item`
   function getValueForAllLevels(level) {
     return function(item) {
-      return item.threadResults[level.idx][cfg.valueAttr]
+      return item.threadResults.hasOwnProperty(level.idx) ? item.threadResults[level.idx][cfg.valueAttr] : null
```
Comment (unrealsolver) [inline public/static/plugins/perf/js/trend_chart.js]: In my opinion data normalization is better approach. This code is ok too, however I would ask you to reduce line length. I prefer
```
cond
  ? A
  : B
```
Also, do you have any reason to use `hasOwnProperty`? Usually, `if (item.threadResults[level.idx]) { ` should be enough.

File: public/static/plugins/perf/js/trend_chart.js (hunk 1 above)
Comment (jimoleary) [reply public/static/plugins/perf/js/trend_chart.js]: You are correct, this one is not needed.

File: public/static/plugins/perf/js/trend_chart.js (hunk 2 above)
Comment (jimoleary) [reply public/static/plugins/perf/js/trend_chart.js]: The thread levels could change in the future, so normalization might be an issue.

I think I used  hasOwnProperty as a cut and paste from the previous case (where 0 would have been false) so the shorter form should be enough.

**Comment by unrealsolver** [comment]
CoffeeScript has `?` syntax, like `item.threadResults[level.idx]?[cfg.valueAttr]` which is neat in such situations. I'm not sure if JS/ES 6/7/next/whatever supports it.

**Pull request merged**